accepted/cranelift-isel-isle-peepmatic.md

```diff
@@ -0,0 +1,926 @@
+
+Second, the DSL encourages a mode of thinking about instructions (both
+in the IR and at the machine level, in VCode) that is more
+value-oriented. The `LowerCtx` API is designed to be a sort of general
+query API, and instruction output registers are just attributes of an
+instruction like any other. Similarly, allocating temporary registers
+and emitting instructions are just imperative actions taken by
+lowering code; there is no notion of the emitted instructions being
+"equal" to values on the input side. In contrast, the ISLE DSL design
+privileges the notion of an instruction's "value". This makes the
+notation more natural when expressing algebraic-style reductions, and
+is consistent with many other instruction-selection systems. (As a
+particular implementation note, multiple-output instructions should
+work within this framework as well because one can treat the N-output
+tuple as a value with constructors and extractors as necessary.)
+However, ISLE does not go as far as an instruction-selector framework
+that is explicitly aware of, e.g., SSA values: rather, that is up to
+the constructors that have been defined. The ISLE system itself only
+understands terms and trees of terms; this just happens to be a
+natural abstraction with which to represent expression trees.
+
+# Discussion: Future Implications for Verification
+
+Though we have not yet worked out all the details, we are confident
+that the translation of rules expressed in the ISLE DSL into some
+machine-readable form for formal verification efforts should be
+possible. This is primarily because of the "equivalent-value"
+semantics that are inherent in a term-rewriting system. The
+denotational value of a term is the symbolic or concrete value
+produced by the instruction it represents (depending on the
+interpretation); so "all" we have to do is to write, e.g.,
+pre/post-conditions for some SMT-solver or theorem-prover that
+describe the semantics of instruction terms on either side of the
+translation.
+
+Note that while externally-defined extractors and constructors at
+first glance may appear to make this more difficult, because they
+define an "FFI" into imperative code, in actuality we can just treat
+them as axiomatic building blocks. In essence, they are the pieces
+that define the input and output instruction sets, and so are tied to
+the definitions of the formal semantics for these two program
+representations; we would start by formally describing the program
+value represented by the result of an extractor on an instruction,
+and/or the preconditions it implies, then carry through the
+implications of this to intermediate internal terms and finally to the
+tree of constructor calls that build the output instructions.
+
+# Example Sketch
+
+
+```
+    ;; --- "FFI" mapping of input instructions and lowering API ---
+    (type Inst extern (enum
+      (Add  (a Input) (b Input))
+      (Load (addr Input))
+      (Const (val u64))
```

Comment:
It's going to be harder to inspect some characteristics of the given resulting Rust enum: what are its size? which variant is the biggest (and thus determines the whole enum's size)? Reducing the memory footprint of IR data structures has been a fruitful exercise in the past, causing reductions in the number of malloc calls and thus compile-time speedups, and I'm a bit worried that this kind of optimization opportunities, that was very easy to implement in the past, will become unbearable within this framework, except for the framework author(s).

On the other hand, one could argue that any improvement to the generic framework would benefit *all* the backends at once (but that's just a benefit from a generic framework: errors/bugs would be shared too!). Still, it doesn't seem like it's fitting the initial goal of being welcoming and simple for newcomers.

Reply:
Ah, so a key thing here is that this `extern` type decl is referring to an enum that's already defined in Rust code elsewhere. Perhaps we could find a better way of doing it later, but for now the idea is to declare the view of the types that we need to generate Rust code against them, but not to generate the type definitions themselves. So during the migration, we still have the `enum Inst` in `isa/*/inst/mod.rs`, and we can still e.g. write tests against its size and play with its packing.

One could argue that it'd be better to generate both from a single source of truth, and I think that has to be the eventual answer; in that case one could still write a unit-test against `std::mem::size_of::<x64::Inst>()` or whatnot but changing it would mean editing the DSL.

A key factor in approachability IMHO (this may be controversial; please discuss!) could be to include the generated code in the git repo. Aside from implications on build-time above, it would mean that (i) the actual Rust definitions, at the stage in our evolution when they *are* all generated from the DSL, should be easy to find; and (ii) we can add a note to the top of the file saying "autogenerated, please see file.isle" and maybe on the actual enum a comment like "autogenerated from definition Foo in file.isle line 123".

I *do* think that there is an interesting philosophical debate about the definition of "welcoming and simple": building some DSL framework means that if we're careful, the "happy path" of adding a new instruction or a new lowering rule is much easier, and also means that there is less cognitive load / fewer subtle invariants to keep in mind. That's pretty much the driving factor I had in mind at least. But I agree that it can make systemic optimizations, like "make the `Inst` enum smaller", more difficult it it means one has to step up into a meta-level, and it also puts a lot of pressure on the framework to be correct, lest we have difficult-to-find bugs. IMHO the tradeoff is worth it and does make the codebase more welcoming for the most common sort of contribution we anticipate in the future, but... please do discuss!